File: Data Structures & Algorithms/maximum-product-subarray/submission-2.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        pre = suf = 0
        n = len(nums)
        res = nums[0]

        for i in range(n):
            multiplier = 1
            if pre != 0:
                multiplier = pre
            
            pre = nums[i] * multiplier

            multiplier = 1
            if suf != 0:
                multiplier = suf

            suf = nums[n - i - 1] * multiplier
            print(pre, suf)
            res = max(res, max(pre, suf))

        return res

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int s, e, res;

        s = 0, e = nums.size() - 1, res = -1;

        while (s <= e) {
            int m = (s + (e - s)/2);

            if (nums[m] == target) {
                res = m;
                break;
            }

            if (nums[s] <= nums[m]) {
                if (nums[s] <= target && nums[m] >= target) {
                    e = m;
                } else {
                    s = m + 1;
                }
            } else {
                if (nums[m] <= target && nums[e] >= target) {
                    s = m;
                } else {
                    e = m - 1;
                }
            }
        }

        return res;
    }
};
